taichi/transforms/optimize_local_variable.cpp

```diff
@@ -0,0 +1,382 @@
+      stored = true;
+      loaded = true;
+      last_store = nullptr;
+      last_store_valid = false;
+      last_store_loaded = false;
+      last_atomic = nullptr;
+      last_atomic_eliminable = false;
+    } else if (!loop.stored) {
+      // Since the loop does not store the alloca,
+      // we can do store-forwarding.
+      if (loop.loaded && inside_loop != 2 &&
+          ((!stored && inside_loop != 1) || last_store_valid)) {
+        loop = *this;
+        loop.inside_loop = 2;
+        body->accept(&loop);
+      }
+      // And the status about the last store should not be changed.
+    } else {
+      // The loop stores the alloca, and it must be invalid now
+      // as we don't know if the loop is fully executed.
+      last_store = loop.last_store;
+      last_store_valid = false;
+      last_atomic = loop.last_atomic;
+      if (loop.loaded) {
+        // The loop loads the alloca, so we cannot eliminate any stores
+        // or AtomicOpStmts in the loop.
+        last_store_loaded = true;
+        last_atomic_eliminable = false;
+      } else {
+        // The loop stores the alloca but never loads it.
+        last_store_loaded = false;
+        last_atomic_eliminable = true;
+      }
+    }
+  }
+
+  void visit(WhileStmt *stmt) override {
+    TI_ASSERT(stmt->mask == nullptr);
+    visit_loop(stmt->body.get(), false);
+  }
+
+  void visit(RangeForStmt *stmt) override {
+    visit_loop(stmt->body.get(), stmt->loop_var == alloca_stmt);
+  }
+
+  void visit(StructForStmt *stmt) override {
+    bool is_loop_var = false;
+    for (auto &loop_var : stmt->loop_vars) {
+      if (loop_var == alloca_stmt) {
+        is_loop_var = true;
+        break;
+      }
+    }
+    visit_loop(stmt->body.get(), is_loop_var);
+  }
+
+  void run() {
+    Block *block = alloca_stmt->parent;
+    TI_ASSERT(block);
+    int location = block->locate(alloca_stmt);
+    TI_ASSERT(location != -1);
+    for (int i = location + 1; i < (int)block->size(); i++) {
+      block->statements[i]->accept(this);
+    }
+    if (last_store && !last_store_loaded) {
+      // The last store is never loaded.
+      // last_store_valid == false means that it's in an IfStmt.
+      // Eliminate the last store.
+      last_store->parent->erase(last_store);
+      throw IRModified();
+    }
+    if (last_atomic && last_atomic_eliminable) {
+      // The last AtomicOpStmt is never loaded.
+      // last_atomic_valid == false means that it's in an IfStmt.
+      if (irpass::analysis::gather_statements(
+              block,
+              [&](Stmt *stmt) { return stmt->have_operand(last_atomic); })
+              .empty()) {
+        // The last AtomicOpStmt is never used.
+        // Eliminate the last AtomicOpStmt.
+        last_atomic->parent->erase(last_atomic);
+        throw IRModified();
+      }
+    }
+    if (!stored && !loaded) {
+      // Never stored and never loaded.
+      if (irpass::analysis::gather_statements(
```

Comment:
Is this `if` for for loops?

Reply:
This is actually for masks in the future. I've set `stored = loaded = true` for for loops.

Reply:
Shall we remove it or add some comments?

I just double-checked that all tests pass if I remove this `if`.

Reply:
Thanks - either moving the `if` or adding some comments to clarify why we need the `if` here would be helpful.